### ISSUE TITLE
Proposal: use template library for JSON configs

### ISSUE THREAD

**Issue by chorsley** [certtools/intelmq#1026]
I'd like to suggest the addition of template-driven config files in IntelMQ using the Jinja2 library (http://jinja.pocoo.org/). 

Straight JSON configs could still be used as-is, so no existing configs would need to change. This change would provide the additional, optional functionality that if any Jinja2 template directives are found in an IntelMQ config file, they get processed. Environment variables are available in the template as `{{ env.MAIL_SERVER }}`.

This should help with the following issues of config management:

* Using interpolated environment variables, secrets no longer need to be embedded in runtime.conf, and config templates can be safely checked into a repository.
* Config templates can be easily shared as-is between organisations.
* Template loops can be used for repetitive bot stanzas, particularly in pipeline.conf, making them easier to manage.
* Configs can be made modular using includes directives (e.g. create runtime-abusech.conf + runtime-cleanmx.conf, and enable or disable bot families easily in runtime.conf with `{% include runtime-abusech.conf %}`)

Simple PoC code in lib/utils.py looks like:

```
def load_configuration(configuration_filepath: str) -> dict:
    if os.path.exists(configuration_filepath):
        config_path, config_filename = os.path.split(configuration_filepath)

        env = jinja2.Environment(
            loader=jinja2.FileSystemLoader(config_path),
            undefined=jinja2.StrictUndefined)
        template = env.get_template(config_filename)
        config_str = template.render(env=os.environ)

        config = json.loads(config_str)
    else:
        raise ValueError('File not found: %r.' % configuration_filepath)
    return config
```

Before I prepare a refined, tested pull request, I was hoping for any feedback on this idea.

**Comment by chorsley** [comment]
One issue observed: IntelMQ Manager currently accesses the raw configs directly. These will not be valid JSON as they include Jinja2 directives, so it would need to be able to grab the configs after Jinja2 processing.

**Comment by wagner-certat** [comment]
I'm not sure how that fit's to intelmq's principles and especially KISS.

> so it would need to be able to grab the configs after Jinja2 processing.

And after saving the config there, the Jinja2 template is lost and you only have the processed config with the changes.

**Comment by chorsley** [comment]
@wagner-certat good point - I hadn't considered that some will be using IntelMQ manager to write configs. Perhaps the compromise is a separate, simple tool which takes Jinja2-based config templates, and transforms them into IntelMQ JSON configs.

Having Jinja 2 templates is proving very useful for repetitive bot configs (e.g. Shadowserver), grouping bots families into sub-files, and commenting parts of the config in and out as needed. So, I'll likely do something like this for my own purposes, but happy to see if there are ways around IntelMQ Manager integration issues.

**Comment by chorsley** [comment]
An alternative approach I'm now trying. The Jinja templates can live in a separate templates dir, and I'm currently using a CLI tool https://sudo.isl.co/shinto-cli/ to render the templates into the regular static config files in /opt/intelmq/etc. That makes sure everything will still work with IntelMQ Manager, at the cost of needing to ensure the templates regenerate the static config files after any change.

Eventually, this could potentially integrate into intelmqctl or a similar command line tool, which could regenerate the final configs from template on each run.

Sample runtime.conf.j2:

"""
    {% include "runtime_ss.conf" %},
    {# include "runtime_abusech.conf" #}
    {# include "runtime_cleanmx.conf" #}
    {# include "runtime_phishtank.conf" #}
    {# include "runtime_zoneh.conf" #}
"""

`{% ... %}` is active, and `{# ... #}` is commented out, making it easy to turn on and off families of bots in the config.

You can then do {{ INTELMQ_SS_* }} are populated from env vars:

"""
    {%- set feeds = {
            'opendns': {
                'subject_rex': 'Shadowserver Open Resolvers',
                'name': 'Shadowserver Open DNS',
                'feedname': 'DNS-open-resolvers'
            },
            ...
     %}

    {% for feed, fdata in feeds.items() %}
        "shadowserver-{{ feed }}-collector": {
            "parameters": {
                "feed": "Shadowserver",
                "provider": "",
                "folder": "INBOX",
                "http_password": null,
                "http_username": null,
                "mail_host": "{{ INTELMQ_SS_MAIL_SERVER }}",
                "mail_password": "{{ INTELMQ_SS_MAIL_PASS }}",
                "mail_ssl": true,
                "mail_user": "{{ INTELMQ_SS_MAIL_ACCOUNT }}",
                "rate_limit": 60,
                "subject_regex": "{{ fdata.subject_rex }}",
                "url_regex": "https://dl.shadowserver.org/.*\\w",
                "ssl_client_certificate": null
            },
            "logging_level": "DEBUG",
            "group": "Collector",
            "name": "{{ fdata.name }} collector",
            "module": "intelmq.bots.collectors.mail.collector_mail_url",
            "description": "Collect Shadowserver URLs from IMAP mailbox"
        },
        "shadowserver-{{ feed }}-parser": {
            "parameters": {
                "feedname": "{{ fdata.feedname }}",
                "overwrite": true
            },
            "group": "Parser",
            "name": "{{ fdata.name }} parser",
            "module": "intelmq.bots.parsers.shadowserver.parser",
            "description": ""
        }{% if not loop.last %},{% endif %}
    {% endfor %}
"""
